### DISCUSSION TITLE
Non-minor cargo dependency updates fail if other dependency depends on prior version

### DISCUSSION THREAD

**Discussion by davidkna** [renovatebot/renovate#29539] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

After updating with a non-minor dependency, `Cargo.toml` renovate fails to update `Cargo.lock` if a different dependency depends on the prior version. It tries to update the old entry, which would make it incompatible, instead of adding a new `Cargo.lock` entry and keeping the old one.

Thus, the `cargo update` command used by `renovate` fails. Among other sub-commands, `cargo check` will update the lockfile correctly early during its run.

[Reproduction repo](https://github.com/davidkna/repro-renovate-lockfile-update) ([PR](https://github.com/davidkna/repro-renovate-lockfile-update/pull/3))

Relevant log:
```
Command failed: cargo update --config net.git-fetch-with-cli=true --manifest-path Cargo.toml --package windows@0.56.0 --precise 0.57.0
    Updating crates.io index
error: failed to select a version for the requirement `windows = "^0.56"`
candidate versions found which didn't match: 0.57.0
location searched: crates.io index
required by package `tauri-winrt-notification v0.2.1`
    ... which satisfies dependency `tauri-winrt-notification = "=0.2.1"` (locked to 0.2.1) of package `repro-renovate-lockfile-update v0.1.0 (/tmp/renovate/repos/github/davidkna/repro-renovate-lockfile-update)`
perhaps a crate was updated and forgotten to be re-vendored?
```

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
What command(s) could renovate run instead?

**Reply by davidkna** [reply]
`cargo fetch` would likely do the least amount of additional work. As far as I can tell, the relevant function in `cargo` itself is `cargo::ops::resolve_ws` which could also be called directly when interfacing with `cargo` as a library.

**Reply by rarkins** [reply]
Please test it on the same commit and verify the results. If it works then we can adopt it. I have doubts that it's as simple as you describe but would be happy to be surprised.

**Reply by davidkna** [reply]
I already verified that `cargo fetch` updates the lockfile. I pushed the changes the command has made to https://github.com/davidkna/repro-renovate-lockfile-update/commit/ae904dcbbab9617b72410f2ef4dff785aa53c030.

**Reply by rarkins** [reply]
How does cargo fetch know to make only the changes to one dependency that we want? What is the exact command?

**Reply by davidkna** [reply]
I only ran `cargo fetch` without any extra arguments. Per [`resolve_ws`](cargo::ops::resolve_ws) which it calls:

> Resolves all dependencies for the workspace using the previous lock file as a guide if present.
>
> This function will also write the result of resolution as a new lock file (unless it is an ephemeral workspace such as cargo install or cargo package).

It sounds like it will only update the parts of the lockfile it needs to resolve the dependency tree with the changes made to the `Cargo.toml`.

**Reply by rarkins** [reply]
From what I can tell from docs and some experimentation, I think you're right about `cargo fetch`. Converted this to an issue: https://github.com/renovatebot/renovate/issues/29620

**Reply by kpreid** [reply]
Hello; I just started trying out Renovate and ran into this problem. It looks like the issue tracking it is in the wrong state — #29620 is closed as completed, but the actual change #30235 was reverted by #31332. The text of #31332 says “Reopens #29620” but this did not actually modify the issue. Could someone with permission to do so reopen #29620?
